docs/support-for-generic-function-types-in-user-defined-type-functions.md

```diff
@@ -0,0 +1,120 @@
+# Support for Generic Types and Packs in User-Defined Type Functions
+
+## Summary
+
+Add support for generic function types in type functions to allow more types to be accepted by them, including tables and classes which might contain generic function types.
+
+## Motivation
+
+Generic functions are very common in Luau, especially when inferred by the typechecking engine. They are also often present in classes defined by the environment.
+
+Omitting generic types from appearing in type functions limits their use.
+Today, even if there is no direct manipulation of generics, a serialization error appears immediately.
+Ideally, we want use-defined type functions to support any possible type, aside from temporary implementation-detail types of the type solver.
+
+## Design
+
+We will introduce a single 'generic' kind of runtime type for both generic types and generic packs.
+
+While this doesn't match up the typechecking engine representation directly, we avoid having to deal with two different userdata types.
+Being content of a generic type or pack is not known upfront, assigning separate tags is not that different from having an extra property on it.
```

Comment:
```suggestion
Since the value of a generic type or pack is not known upfront, assigning separate tags is not that different from having an extra property on it.
```